File: 2251-number-of-flowers-in-full-bloom/2251-number-of-flowers-in-full-bloom.py
class Solution:
    def fullBloomFlowers(self, flowers: List[List[int]], persons: List[int]) -> List[int]:
        
        timeLine = sorted(list(set([ i for s, e in flowers for i in (s, e)] + [i for i in persons])))
        
        flowers.sort()
        persons = [(v, i) for i, v in enumerate(persons)]
        persons.sort()
        activeHeap = []
        i = 0
        pI = 0
        ans = [0]*len(persons)

        for tl in timeLine:
            
            while i < len(flowers) and tl == flowers[i][0]:
                heapq.heappush(activeHeap, (flowers[i][1], flowers[i][0]))
                i+=1
                
            while pI < len(persons) and tl == persons[pI][0]:
                #print(activeHeap, pI, tl, len(activeHeap))
                time, index = persons[pI]
                #print(time, index)
                ans[index] = len(activeHeap)
                #print(ans)
                pI += 1
            #print(tl, activeHeap)   
            while activeHeap and tl == activeHeap[0][0]:
                heapq.heappop(activeHeap)
        
        
        
        return ans

File: 2251-number-of-flowers-in-full-bloom/2251-number-of-flowers-in-full-bloom.java
class Solution {
    public int[] fullBloomFlowers(int[][] flowers, int[] persons) {
        // differences between a TreeMap and a HashMap (sorted right away)
        
        TreeMap<Integer, Integer> prefixMap = new TreeMap<>();
        int[] ans = new int[persons.length];
        
  
        for (int i=0; i < flowers.length; i++) {
            int prevFlowersStart = prefixMap.getOrDefault(flowers[i][0], 0);
            int prevFlowersEnd = prefixMap.getOrDefault(flowers[i][1]+1, 0);
            prefixMap.put(flowers[i][0], prevFlowersStart + 1);
            prefixMap.put(flowers[i][1]+1, prevFlowersEnd - 1);
            //System.out.println(flowers[i][1]);
 

        };
        TreeMap<Integer, Integer> sumMap = new TreeMap<>();
        int runningSum = 0;
        for (Map.Entry<Integer, Integer> sumFlowers: prefixMap.entrySet() ) {
            runningSum += sumFlowers.getValue();
            sumMap.put(sumFlowers.getKey(), runningSum);
            
        }
        
        //System.out.println(sumMap);
        for (int i = 0; i < persons.length; i++) {
            //System.out.println(sumMap.floorEntry(persons[i]));
            Map.Entry<Integer, Integer> floorKey = sumMap.floorEntry(persons[i]);
            if (floorKey == null) {
                ans[i] = 0;
            } else {
                ans[i] = floorKey.getValue();
            }

        }
        
        
        return ans;   
    }
}